From: Chip Davis <cdavis@codeweavers.com>
Subject: [PATCH v3] ucrtbase: Fix handling of tab and non-breaking space in iswctype().
Message-Id: <20191211232355.70180-1-cdavis@codeweavers.com>
Date: Wed, 11 Dec 2019 17:23:55 -0600

It's just these two that are handled specially AFAICT.

Signed-off-by: Chip Davis <cdavis@codeweavers.com>
---

Notes:
    v2: Fix build.
    v3: Fix sign extension bug.

 dlls/msvcr110/tests/msvcr110.c | 30 ++++++++++++++++++++++++++++++
 dlls/msvcrt/tests/string.c     | 28 ++++++++++++++++++++++++++++
 dlls/ucrtbase/tests/misc.c     | 29 +++++++++++++++++++++++++++++
 3 files changed, 87 insertions(+)

diff --git a/dlls/msvcr110/tests/msvcr110.c b/dlls/msvcr110/tests/msvcr110.c
index 11111111111..11111111111 100644
--- a/dlls/msvcr110/tests/msvcr110.c
+++ b/dlls/msvcr110/tests/msvcr110.c
@@ -43,6 +43,7 @@ typedef struct {
 
 static char* (CDECL *p_setlocale)(int category, const char* locale);
 static size_t (CDECL *p___strncnt)(const char *str, size_t count);
+static int (CDECL *p_iswctype)(wchar_t, wctype_t);
 
 static unsigned int (CDECL *p_CurrentScheduler_GetNumberOfVirtualProcessors)(void);
 static unsigned int (CDECL *p__CurrentScheduler__GetNumberOfVirtualProcessors)(void);
@@ -68,6 +69,7 @@ static BOOL init(void)
 
     SET(p_setlocale, "setlocale");
     SET(p___strncnt, "__strncnt");
+    SET(p_iswctype, "iswctype");
     SET(p_CurrentScheduler_GetNumberOfVirtualProcessors, "?GetNumberOfVirtualProcessors@CurrentScheduler@Concurrency@@SAIXZ");
     SET(p__CurrentScheduler__GetNumberOfVirtualProcessors, "?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ");
     SET(p_CurrentScheduler_Id, "?Id@CurrentScheduler@Concurrency@@SAIXZ");
@@ -170,6 +172,33 @@ static void test___strncnt(void)
     }
 }
 
+static void test_iswctype(void)
+{
+    static const struct {
+        WCHAR c;
+        int t;
+        int r;
+    } tests[] = {
+        { '0', C1_DIGIT, C1_DIGIT },
+        { '9', C1_DIGIT, C1_DIGIT },
+        { 'a', C1_DIGIT, 0 },
+        { 'a', C1_LOWER, C1_LOWER },
+        { 0xa0, C1_BLANK, C1_BLANK },
+        { 0xe0, _ALPHA, C1_ALPHA|C1_LOWER },
+        { 0xff16, C1_DIGIT, C1_DIGIT },
+        { 0x0660, C1_DIGIT, C1_DIGIT },
+        { 0x0ce6, C1_DIGIT, C1_DIGIT }
+    };
+    int i, r;
+
+    p_setlocale(LC_ALL, "C");
+    for (i = 0; i < ARRAY_SIZE(tests); i++)
+    {
+        r = p_iswctype(tests[i].c, tests[i].t);
+        ok(r == tests[i].r, "iswctype returned %x for %x\n", r, tests[i].c);
+    }
+}
+
 static void test_CurrentContext(void)
 {
     _Context _ctx, *ret;
@@ -190,5 +219,6 @@ START_TEST(msvcr110)
     test_CurrentScheduler(); /* MUST be first (at least among Concurrency tests) */
     test_setlocale();
     test___strncnt();
+    test_iswctype();
     test_CurrentContext();
 }
diff --git a/dlls/msvcrt/tests/string.c b/dlls/msvcrt/tests/string.c
index 11111111111..11111111111 100644
--- a/dlls/msvcrt/tests/string.c
+++ b/dlls/msvcrt/tests/string.c
@@ -4477,6 +4477,33 @@ static void test_wcsncpy(void)
             wine_dbgstr_wn(dst, ARRAY_SIZE(dst)));
 }
 
+static void test_iswctype(void)
+{
+    static const struct {
+        WCHAR c;
+        int t;
+        int r;
+    } tests[] = {
+        { '0', C1_DIGIT, C1_DIGIT },
+        { '9', C1_DIGIT, C1_DIGIT },
+        { 'a', C1_DIGIT, 0 },
+        { 'a', C1_LOWER, C1_LOWER },
+        { 0xa0, C1_BLANK, C1_BLANK },
+        { 0xe0, _ALPHA, C1_ALPHA|C1_LOWER },
+        { 0xff16, C1_DIGIT, C1_DIGIT },
+        { 0x0660, C1_DIGIT, C1_DIGIT },
+        { 0x0ce6, C1_DIGIT, C1_DIGIT }
+    };
+    int i, r;
+
+    setlocale(LC_ALL, "C");
+    for (i = 0; i < ARRAY_SIZE(tests); i++)
+    {
+        r = iswctype(tests[i].c, tests[i].t);
+        ok(r == tests[i].r, "iswctype returned %x for %x\n", r, tests[i].c);
+    }
+}
+
 START_TEST(string)
 {
     char mem[100];
@@ -4633,4 +4660,5 @@ START_TEST(string)
     test_SpecialCasing();
     test__mbbtype();
     test_wcsncpy();
+    test_iswctype();
 }
diff --git a/dlls/ucrtbase/tests/misc.c b/dlls/ucrtbase/tests/misc.c
index 11111111111..11111111111 100644
--- a/dlls/ucrtbase/tests/misc.c
+++ b/dlls/ucrtbase/tests/misc.c
@@ -36,6 +36,7 @@
 
 #include <windef.h>
 #include <winbase.h>
+#include <winnls.h>
 #include "wine/test.h"
 
 #define DEFINE_EXPECT(func) \
@@ -574,6 +575,33 @@ static void test_isblank(void)
     }
 }
 
+static void test_iswctype(void)
+{
+    static const struct {
+        WCHAR c;
+        int t;
+        int r;
+    } tests[] = {
+        { '0', C1_DIGIT, C1_DIGIT },
+        { '9', C1_DIGIT, C1_DIGIT },
+        { 'a', C1_DIGIT, 0 },
+        { 'a', C1_LOWER, C1_LOWER },
+        { 0xa0, C1_BLANK, 0 },
+        { 0xe0, _ALPHA, C1_ALPHA|C1_LOWER },
+        { 0xff16, C1_DIGIT, C1_DIGIT },
+        { 0x0660, C1_DIGIT, C1_DIGIT },
+        { 0x0ce6, C1_DIGIT, C1_DIGIT }
+    };
+    int i, r;
+
+    _setlocale(LC_ALL, "C");
+    for (i = 0; i < ARRAY_SIZE(tests); i++)
+    {
+        r = _iswctype(tests[i].c, tests[i].t);
+        ok(r == tests[i].r, "iswctype returned %x for %x\n", r, tests[i].c);
+    }
+}
+
 static struct _exception exception;
 
 static int CDECL matherr_callback(struct _exception *e)
@@ -1622,6 +1650,7 @@ START_TEST(misc)
     test__sopen_s();
     test_lldiv();
     test_isblank();
+    test_iswctype();
     test_math_errors();
     test_asctime();
     test_strftime();
